### PR TITLE
if allow_patient_portal is YES set portal login user name if necessary

### LINKED ISSUE #7643: bug:patient's portal login user name can get stuck as empty
### Describe the bug
if a patient's portal login user name is empty, for some reason, it can't be set again, so the user is stuck not being able to use the patient portal
although it's not clear how this can happen, it shouldn't be possible via the user interfce, if it _does_  happen then there's no way round it other than adding in a login user name via the db, which is not advisable

### To Reproduce
 set the login user name to empty in the db, in order to test/demo

edit the demographics for that patient, set 'allow use of patient portal' to 'yes' 
check on the Patient Portal card, reset credentials form - the login user name is still blank

### Expected behavior
if allowed to use the portal is set to 'yes', the login user name should be generated if it doesn't yet exist

### Client configuration

- Browser: firefox
- OpenEMR version: 7.0.3-dev
- Operating system: ubuntu

NOTE - ruthkonyn is working on this

### PR DESCRIPTION
**Pull request by ruthkonyn**
if allow_patient_portal is YES check 'portal login user name' - if not already set - set it to 'portal user name' - otherwise it can get stuck empty, i.e. one can't give it a value.

Fixes #7643

#### Short description of what this resolves:
if portal login user name is empty it can get stuck with no value, and no way to enter a value, under some undefined conditions. 

#### Changes proposed in this pull request:
when the demographic form is saved, check if 'allow patient portal' has the value 'YES', and if so check that 'portal login user name' has a value and if not set it to the same string as 'portal user name'

### CODE REVIEW

**Comment by ruthkonyn** [comment]
i hope this is the appropriate place to put this check and piece of code?

File: src/Services/PatientService.php [hunk 1]
```diff
@@ -245,7 +245,20 @@ public function databaseUpdate($data)
             // need to save off our care team
             $this->saveCareTeamHistory($data, $dataBeforeUpdate['care_team_provider'], $dataBeforeUpdate['care_team_facility']);
         }
-
```
Comment (kojiromike) [inline src/Services/PatientService.php]: This loose comparison could match unexpected values (e.g., `$data['allow_patient_portal'] == true` would match). Also, is `allow_patient_portal` guaranteed to be present in `$data`? If the demographics save doesn't include this field, this would emit a PHP warning.

```suggestion
        if (($data['allow_patient_portal'] ?? '') === 'YES') {
```

File: src/Services/PatientService.php [hunk 2]
```diff
@@ -245,7 +245,20 @@ public function databaseUpdate($data)
...
+        // RM if  'allow_patient_portal' is YES, make sure portal login username has been created
+        if ($data['allow_patient_portal'] == 'YES') {
+            // we're about to set it to YES, so make sure credentials have been created
+            $sql = "SELECT portal_login_username, portal_username FROM patient_access_onsite WHERE pid = ?";
```
Comment (kojiromike) [inline src/Services/PatientService.php]: `sqlFetchArray` returns `false` when no row exists in `patient_access_onsite` for this patient (e.g., credentials were never created). Accessing `$names['portal_login_username']` on `false` would cause a PHP warning. Could you add a guard here?

```suggestion
            $names = sqlFetchArray($sqlget);
            if ($names !== false && ($names['portal_login_username'] ?? '') === '') {
```

(This would replace the separate `if` on the next line.)

File: src/Services/PatientService.php [hunk 3]
```diff
@@ -245,7 +245,20 @@ public function databaseUpdate($data)
...
+            $sqlget = sqlStatement($sql, $data['pid']);
+            $names = sqlFetchArray($sqlget);
+            if ($names['portal_login_username'] == "") {
```
Comment (kojiromike) [inline src/Services/PatientService.php]: The SQL uses positional `?` placeholders, but below you're building an associative array with named keys. While this may work in practice (PDO iterates values in insertion order), it's fragile and inconsistent with the rest of the codebase. Could you simplify to a plain indexed array?

```suggestion
                $sql = "UPDATE patient_access_onsite SET portal_login_username = ? WHERE pid = ?";
                sqlStatement($sql, [$names['portal_username'], $data['pid']]);
```

(This also removes the unused `$sqlres` and `$update_parameters` variables.)

File: src/Services/PatientService.php [hunk 4]
```diff
@@ -245,7 +245,20 @@ public function databaseUpdate($data)
             // need to save off our care team
             $this->saveCareTeamHistory($data, $dataBeforeUpdate['care_team_provider'], $dataBeforeUpdate['care_team_facility']);
         }
```
Comment (kojiromike) [inline src/Services/PatientService.php]: Minor style note: the `RM` initials in comments aren't a convention used elsewhere in the codebase. Could you rephrase as a plain description of what the code does?

```suggestion
        // If portal access is enabled, ensure the portal login username exists
```

**Review by kojiromike — CHANGES_REQUESTED** [review]
Thanks for working on this — the intent is good and this is a real usability problem. A few things to address before this can merge:

**Comment by ruthkonyn** [comment]
@kojiromike  - thanks for looking at this so it can finally be used - I should have time this week to resolve your comments.

**Comment by ruthkonyn** [comment]
@kojiromike - needless to say after i'd implemented your suggested changes and i did a phpstan reivew, i got a handful or two of phpstan error messages, mainly about existing lines of code that i hadn't changed myself. I'm working through these, but have to fit this work round other commitments that i have. I assume it's not urgent though...

**Comment by ruthkonyn** [comment]
hi @kojiromike - i've done as much as i can - i don't know how to fix the phpstan and Conventinal Commits errors - would you like to suggest changes for these, and anything else you think needs changes, thanks

File: src/Services/PatientService.php (hunk 2 above)
Comment (kojiromike) [reply src/Services/PatientService.php]: Addressed in the cleanup commit — now uses `QueryUtils::querySingleRow()` with `$names !== false` guard.

File: src/Services/PatientService.php (hunk 3 above)
Comment (kojiromike) [reply src/Services/PatientService.php]: Addressed in the cleanup commit — uses plain indexed array `[$names['portal_username'], $data['pid']]` with `QueryUtils::sqlStatementThrowException()`.

File: src/Services/PatientService.php (hunk 4 above)
Comment (kojiromike) [reply src/Services/PatientService.php]: Addressed in the cleanup commit — comment now reads `// If portal access is enabled, ensure the portal login username exists`.

File: src/Services/PatientService.php [hunk 5]
```diff
@@ -276,6 +276,16 @@ public function databaseUpdate($data)
...
+        if (is_array($data) && ($data['allow_patient_portal'] ?? '') === 'YES') {
+            $portalSql = "SELECT portal_login_username, portal_username FROM patient_access_onsite WHERE pid = ?";
+            $names = QueryUtils::querySingleRow($portalSql, [$data['pid']]);
+            if ($names !== false && ($names['portal_login_username'] ?? '') === '') {
+                $updateSql = "UPDATE patient_access_onsite SET portal_login_username = ? WHERE pid = ?";
+                QueryUtils::sqlStatementThrowException($updateSql, [$names['portal_username'], $data['pid']]);
+            }
+        }
+
         if ($sqlResult) {
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply src/Services/PatientService.php]: Good catch — moved the portal username repair block inside the `if ($sqlResult)` check so it only runs after the main patient_data UPDATE succeeds. No need for a try/catch — if the portal update fails, it should propagate rather than silently log and continue.

File: src/Services/PatientService.php [hunk 6]
```diff
@@ -276,6 +276,16 @@ public function databaseUpdate($data)
         array_push($query['bind'], $data['pid']);
         $sqlResult = sqlStatement($sql, $query['bind']);
 
+        // If portal access is enabled, ensure the portal login username exists
+        if (is_array($data) && ($data['allow_patient_portal'] ?? '') === 'YES') {
+            $portalSql = "SELECT portal_login_username, portal_username FROM patient_access_onsite WHERE pid = ?";
+            $names = QueryUtils::querySingleRow($portalSql, [$data['pid']]);
+            if ($names !== false && ($names['portal_login_username'] ?? '') === '') {
+                $updateSql = "UPDATE patient_access_onsite SET portal_login_username = ? WHERE pid = ?";
+                QueryUtils::sqlStatementThrowException($updateSql, [$names['portal_username'], $data['pid']]);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply src/Services/PatientService.php]: `sqlStatementThrowException` is intentional here — it's the modern QueryUtils pattern for writes. If the portal update fails due to a DB error, that should propagate as an exception rather than be silently swallowed. The method doesn't catch exceptions elsewhere either, and callers are expected to handle them.

File: src/Services/PatientService.php [hunk 7]
```diff
@@ -276,6 +276,16 @@ public function databaseUpdate($data)
         array_push($query['bind'], $data['pid']);
         $sqlResult = sqlStatement($sql, $query['bind']);
 
+        // If portal access is enabled, ensure the portal login username exists
+        if (is_array($data) && ($data['allow_patient_portal'] ?? '') === 'YES') {
+            $portalSql = "SELECT portal_login_username, portal_username FROM patient_access_onsite WHERE pid = ?";
+            $names = QueryUtils::querySingleRow($portalSql, [$data['pid']]);
+            if ($names !== false && ($names['portal_login_username'] ?? '') === '') {
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply src/Services/PatientService.php]: Agreed that test coverage would be valuable, but it's out of scope for this contributor's bug fix — it would require database setup and `patient_access_onsite` fixtures. The existing PatientService tests don't cover this table.

File: src/Services/PatientService.php [hunk 8]
```diff
@@ -277,6 +277,15 @@ public function databaseUpdate($data)
         $sqlResult = sqlStatement($sql, $query['bind']);
 
         if ($sqlResult) {
+            // If portal access is enabled, ensure the portal login username exists
+            if (is_array($data) && ($data['allow_patient_portal'] ?? '') === 'YES') {
+                $portalSql = "SELECT portal_login_username, portal_username FROM patient_access_onsite WHERE pid = ?";
+                $names = QueryUtils::querySingleRow($portalSql, [$data['pid']]);
+                if ($names !== false && ($names['portal_login_username'] ?? '') === '') {
+                    $updateSql = "UPDATE patient_access_onsite SET portal_login_username = ? WHERE pid = ?";
+                    QueryUtils::sqlStatementThrowException($updateSql, [$names['portal_username'], $data['pid']]);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply src/Services/PatientService.php:286]: The suggested `updatePortalLoginUsername` method doesn't exist on `PatientAccessOnsiteService`. The service's credential update flow (`saveCredentials`) dispatches `PortalCredentialsUpdatedEvent` but requires a password parameter — it's designed for full credential creation/reset, not for backfilling an empty login username.

This fix is a narrow repair: when `portal_login_username` is empty but `portal_username` exists, copy one to the other. It doesn't change the password or the actual username — it fills in a missing derived field. Routing through the full credential update flow would be incorrect here since we don't have (and shouldn't need) a password.

If modules need to observe this specific repair, that would be a separate enhancement.

File: tests/Tests/Services/PatientServiceTest.php [hunk 9]
```diff
@@ -126,6 +127,36 @@ public function testUpdateSuccess(): void
...
+
+        // Create portal credentials with a username but empty login username
+        QueryUtils::sqlStatementThrowException(
+            "INSERT INTO patient_access_onsite (pid, portal_username, portal_login_username) VALUES (?, ?, '')",
+            [$pid, 'testportaluser']
+        );
+
+        // Update patient with portal access enabled — should backfill login username
+        $this->patientFixture['allow_patient_portal'] = 'YES';
+        $this->patientService->update($actualUuid, $this->patientFixture);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply tests/Tests/Services/PatientServiceTest.php]: Already fixed in the previous commit — the test now calls `databaseUpdate()` directly.

File: tests/Tests/Services/PatientServiceTest.php [hunk 10]
```diff
@@ -126,6 +127,36 @@ public function testUpdateSuccess(): void
...
+
+        $row = QueryUtils::querySingleRow("SELECT portal_login_username FROM patient_access_onsite WHERE pid = ?", [$pid]);
+        $this->assertIsArray($row);
+        $this->assertSame('testportaluser', $row['portal_login_username']);
+
+        // Clean up
+        QueryUtils::sqlStatementThrowException("DELETE FROM patient_access_onsite WHERE pid = ?", [$pid]);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply tests/Tests/Services/PatientServiceTest.php]: Good catch — wrapped the test body in try/finally so the `patient_access_onsite` cleanup runs even if assertions fail.

File: tests/Tests/Services/PatientServiceTest.php [hunk 11]
```diff
@@ -126,6 +127,39 @@ public function testUpdateSuccess(): void
...
+    public function testUpdateBackfillsPortalLoginUsername(): void
+    {
+        $actualResult = $this->patientService->insert($this->patientFixture);
+        $this->assertTrue($actualResult->isValid());
+        $data = $actualResult->getData();
+        $this->assertIsArray($data);
+        $dataResult = $data[0];
+        $this->assertIsArray($dataResult);
+        $pid = $dataResult['pid'];
+        $actualUuid = $dataResult['uuid'];
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply tests/Tests/Services/PatientServiceTest.php]: Good catch — removed. It was left over from when the test called `update()` (uuid-based) instead of `databaseUpdate()` (pid-based).

File: src/Services/PatientService.php [hunk 12]
```diff
@@ -277,6 +277,15 @@ public function databaseUpdate($data)
         $sqlResult = sqlStatement($sql, $query['bind']);
 
         if ($sqlResult) {
+            // If portal access is enabled, ensure the portal login username exists
+            if (is_array($data) && ($data['allow_patient_portal'] ?? '') === 'YES') {
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply src/Services/PatientService.php:281]: The PR description says "when enabling portal access" but the underlying bug (#7643) is that `portal_login_username` can get stuck empty under undefined conditions — not only during the initial enable. Running the check on every save where portal is YES ensures the repair catches all cases. If the username is already populated, the inner `if` short-circuits without issuing an UPDATE, so the cost is just one lightweight SELECT on saves where portal is enabled.

**Comment by ruthkonyn** [comment]
many thanks @kojiromike
